Translated 'Rendering Lists' page (#670)
reactjs/pt-br.react.dev: src/content/learn/rendering-lists.md

---
title: Renderizando Listas
---

<Intro>

> Você frequentemente gostará de exibir múltiplos componentes similares a partir de uma coleção de dados. Você pode usar os [métodos de array JavaScript](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array#) para manipular um array de dados. Nesta página, você usará [`filter()`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) e [`map()`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/map) com o React para filtrar e transformar seu array de dados em um array de componentes.

Reviewer: Tu pode usar as traduções do site do mozilla.
https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/filter
https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/map

</ul>
```

> As únicas diferenças entre os itens da lista são seus conteúdos, seus dados. Você comumente precisará exibir diversas instâncias do mesmo componente usando diferentes dados ao construir interfaces: de listas de comentários a galeiras de fotos de perfil. Nessas situações, você pode armazenar esses dados em objetos ou arrays JavaScript e usar métodos como [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) e [`filter()`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) para renderizar listas de componentes a partir deles.

Reviewer 2: ```suggestion
As únicas diferenças entre os itens da lista são seus conteúdos, seus dados. Você comumente precisará exibir diversas instâncias do mesmo componente usando diferentes dados ao construir interfaces: de listas de comentários a galerias de fotos de perfil. Nessas situações, você pode armazenar esses dados em objetos ou arrays JavaScript e usar métodos como [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) e [`filter()`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) para renderizar listas de componentes a partir deles.
```

> As únicas diferenças entre os itens da lista são seus conteúdos, seus dados. Você comumente precisará exibir diversas instâncias do mesmo componente usando diferentes dados ao construir interfaces: de listas de comentários a galerias de fotos de perfil. Nessas situações, você pode armazenar esses dados em objetos ou arrays JavaScript e usar métodos como [`map()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) e [`filter()`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) para renderizar listas de componentes a partir deles.

Reviewer 3: ```suggestion
As únicas diferenças entre os itens da lista são seus conteúdos, seus dados. Você comumente precisará exibir diversas instâncias do mesmo componente usando diferentes dados ao construir interfaces: de listas de comentários a galerias de fotos de perfil. Nessas situações, você pode armazenar esses dados em objetos ou arrays JavaScript e usar métodos como [`map()`](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/map) e [`filter()`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) para renderizar listas de componentes a partir deles.
```

</Sandpack>

Perceba que a sandbox acima exibe um erro no console:

<ConsoleBlock level="error">

> Warning: Each child in a list should have a unique "key" prop.

Reviewer 3: Da para traduzir esse trecho

}];
```

Digamos que você queira uma maneira de mostrar apenas as pessoas cuja profissão é `'chemist'`. Você pode usar o método `filter()` do JavaScript para retornar apenas essas pessoas. Esse método recebe um array de itens, passa-os por um “teste” (uma função que retorna `true` ou `false`), e retorna um novo array contendo apenas aqueles itens os quais passaram no teste (retornaram `true`).

> Você quer apenas os itens onde `profession` seja `'chemist'`. A função "teste" para isso se parece com `(person) => person.profession === 'chemist'`. Aqui está como juntar tudo isso:

Author: Gostei das correções porém o `. Profession` ali está errado 😅

> 1. **Crie** um novo array de apenas pessoas “chemist”, `chemists`, chamando `filter()` em `people` e filtrando por `person.profession === 'chemist'`:

Author: Acho importante manter a relação com a variável `chemists` feita depois da explicação sobre o que ela contém

```js
const chemists = people.filter(person =>
  person.profession === 'chemist'
);
```

> 2. Agora **mapeie** sobre `chemists`:

Author: Com base na [definição do método na MDN](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Array/map) e por se tratar de um termo da área acredito que seja melhor deixar como está.

<Pitfall>

Arrow functions retornam implicitamente a expressão logo após `=>`, então você não precisa de uma declaração `return`:

```js
const listItems = chemists.map(person =>
  <li>...</li> // Retorno implícito!
);
```

Entretanto, **você deve escrever `return` explicitamente se seu `=>` é seguido de uma chave `{`!**

```js
const listItems = chemists.map(person => { // Chave
  return <li>...</li>;
});
```

> Arrow functions contendo `=> {` são ditas como havendo um ["corpo de bloco".](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#function_body) Elas deixam com que você escreva mais de uma linha de código, mas você *precisa* escrever uma declaração `return` você mesmo. Caso esquecida, nada é retornado!

Reviewer 3: ```suggestion
Arrow functions contendo `=> {` são ditas como havendo um ["corpo de bloco".](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Functions/Arrow_functions#function_body) Elas deixam com que você escreva mais de uma linha de código, mas você *precisa* escrever uma declaração `return` você mesmo. Caso esquecida, nada é retornado!
```

</Pitfall>

## Mantendo itens em ordem com `key` {/*keeping-list-items-in-order-with-key*/}

Perceba que todas as sandboxes acima exibem um erro no console:

<ConsoleBlock level="error">

> Warning: Each child in a list should have a unique "key" prop.

Reviewer 3: aqui também

> Aviso: Cada filho em uma lista deve haver uma prop "key" única.

Author: Hmmm, não vejo necessidade do warning ser traduzido de uma maneira tão verbosa, no meu ver os termos cada e única já dão a conotação necessária.
Única alteração que fiz foi trocar "haver" para "ter"

</ConsoleBlock>

> Você precisa dar a cada item do array uma `key` -- uma string ou um número o qual o identifica unicamente dentre os demais itens naquele array:

Author: Acho que conjugação da sua sugestão está errada, na frase "uma string ou um número o qual o identifica unicamente" nos referimos ao termo "item", no singular. 
Dito isto, o entendimento pode ser melhorado reescrevendo dessa forma: "uma string ou um número que o identifique unicamente"

```js
<li key={person.id}>...</li>
```

<Note>

> Elementos JSX diretamente dentro de um chamado `map()` sempre precisam de keys!

Author: Que tal "Elementos JSX retornados por um chamado `map()` sempre precisam de keys!"?

> Elementos JSX diretamente dentro de um chamado `map()` sempre precisam de keys!

Reviewer 4: Que tal substituir a palavra "chamado" por sua forma feminina? Eu percebo que a comunidade tende a utilizar mais a palavra "chamada" no gênero feminino.

);
```

Fragmentos desaparecem da DOM, então isto irá produzir uma lista plana de `<h1>`, `<p>`, `<h1>`, `<p>`, e assim por diante.

</DeepDive>

### Onde conseguir sua `key` {/*where-to-get-your-key*/}

Diferentes fontes de dados fornecem diferentes fontes de keys:

> * **Dados de uma base de dados:** Se seus dados estão vindo de uma base de dados, você pode usar as keys/IDs desta, os quais são únicos por natureza.

Author: Acredito que dessa forma a leitura fique um pouco repetitiva visto que no início da frase o termo "dado" já é bastante citado e apenas "base" soe um pouco estranho

> * **Dados gerados localmente:** Se seus dados são gerados e persistidos localmente (ex. anotações em um aplicativo de notas), use um contador incremental, [`crypto.randomUUID()`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID) ou um pacote como [`uuid`](https://www.npmjs.com/package/uuid) ao criar itens.

Reviewer 3: ```suggestion
* **Dados gerados localmente:** Se seus dados são gerados e persistidos localmente (ex. anotações em um aplicativo de notas), use um contador incremental, [`crypto.randomUUID()`](https://developer.mozilla.org/pt-BR/docs/Web/API/Crypto/randomUUID) ou um pacote como [`uuid`](https://www.npmjs.com/package/uuid) ao criar itens.
```

</Solution>

> #### Listas aninhadas em um componente {/*nested-lists-in-one-component*/}

Author: Encadeado dá o sentido de que uma lista está após a outra, não em seu interior.

</Sandpack>

Aqui, `<Recipe {...recipe} key={recipe.id} />` é um atalho de sintaxe dizendo "passe todas as propriedades do objeto `recipe` como props para o componente `Recipe`". Você também pode escrever cada prop explicitamente: `<Recipe id={recipe.id} name={recipe.name} ingredients={recipe.ingredients} key={recipe.id} />`.

> **Note que a `key` é especificada no `<Recipe>` propriamente dito ao invés de na `div` raiz retornada por `Recipe`.** Isso se deve ao fato de que essa `key` é diretamente necessária dentro do contexto do array circundante. Anteriormente, você tinha um array de `<div>`s então cada uma delas precisava de uma `key`, mas agora você tem um array de `<Recipe>`s. Em outras palavras, quando você extrai um componente, não se esqueça de deixar a `key` fora da JSX que você copia e cola.

Reviewer 4: Eu acho que a palavra "circundante" não fica muito claro, acho melhor substituir por "em que o componente está sendo renderizado"

```suggestion
**Note que a `key` é especificada no `<Recipe>` propriamente dito ao invés de na `div` raiz retornada por `Recipe`.** Isso se deve ao fato de que essa `key` é diretamente necessária dentro do contexto do array em que o componente está sendo renderizado. Anteriormente, você tinha um array de `<div>`s então cada uma delas precisava de uma `key`, mas agora você tem um array de `<Recipe>`s. Em outras palavras, quando você extrai um componente, não se esqueça de deixar a `key` fora da JSX que você copia e cola.
```

</Sandpack>

> Lembre-se, fragmentos (comumente escritos como `<> </>`) deixam com que você agrupe nós JSX sem adicionar `<div>`s extras!

Reviewer 4: ```suggestion
Lembre-se, fragmentos (são comumente escritos como `<> </>`) deixam com que você agrupe nós JSX sem adicionar `<div>`s extras!
```

Author: Acredito que isso dificulte o entendimento da frase, no entanto o restante das sugestões é bastante válido!
